kythe/cxx/verifier/souffle_entrysteam_support.cc

```diff
@@ -0,0 +1,159 @@
+#include "google/protobuf/io/zero_copy_stream_impl.h"
+#include "kythe/proto/storage.pb.h"
+#include "souffle/io/IOSystem.h"
+
+#ifdef __APPLE__
+extern "C" {
+// TODO(zarko): fix these on darwin
+void ffi_call() { abort(); }
+void ffi_prep_cif_machdep() { abort(); }
+void ffi_prep_closure_loc() { abort(); }
+}
+#endif  // defined(__APPLE__)
+
+namespace kythe {
+namespace {
+
+/// Number of elements in the vname representation and their offsets.
+constexpr int kVNameElements = 5;
+constexpr int kSignatureEntry = 0;
+constexpr int kCorpusEntry = 1;
+constexpr int kPathEntry = 2;
+constexpr int kRootEntry = 3;
+constexpr int kLanguageEntry = 4;
+/// Number of elements in the entry representation and their offsets.
+constexpr int kEntryElements = 4;
+constexpr int kSourceEntry = 0;
+constexpr int kKindEntry = 1;
+constexpr int kTargetEntry = 2;
+constexpr int kValueEntry = 3;
+
+class KytheEntryStreamReadStream : public souffle::ReadStream {
+ public:
+  explicit KytheEntryStreamReadStream(
+      const std::map<std::string, std::string>& rw_operation,
+      souffle::SymbolTable& symbol_table, souffle::RecordTable& record_table)
+      : souffle::ReadStream(rw_operation, symbol_table, record_table),
+        raw_input_(STDIN_FILENO) {
+    // Check that the relation has the correct type.
+    if (typeAttributes.size() == kEntryElements &&
+        typeAttributes[0] == "r:vname" && typeAttributes[1] == "s:symbol" &&
+        typeAttributes[2] == "r:vname" && typeAttributes[3] == "s:symbol") {
+      const auto& record_info = types["records"]["r:vname"];
+      size_t arity = record_info["arity"].long_value();
+      const auto& types = record_info["types"];
+      if (arity == kVNameElements) {
+        relation_ok_ = true;
+        for (size_t i = 0; i < arity; ++i) {
+          if (types[i].string_value() != "s:symbol") {
+            relation_ok_ = false;
+            break;
+          }
+        }
+      }
+    }
+    if (!relation_ok_) {
+      fprintf(stderr, R"(error: bad relation for kythe_entrystream. Use:
+
+.type vname = [
+  signature:symbol,
+  corpus:symbol,
+  path:symbol,
+  root:symbol,
+  language:symbol
+]
+
+.decl entry(source:vname, kind:symbol, target:vname, value:symbol)
+
+Note that this is an experimental feature and this declaration may change.
+)");
+    }
+  }
+
+ protected:
+  souffle::Own<souffle::RamDomain[]> readNextTuple() override {
+    if (!relation_ok_) {
+      return nullptr;
+    }
+    google::protobuf::uint32 byte_size;
+    google::protobuf::io::CodedInputStream coded_input(&raw_input_);
+    coded_input.SetTotalBytesLimit(INT_MAX);
+    if (!coded_input.ReadVarint32(&byte_size)) {
+      return nullptr;
+    }
+    coded_input.PushLimit(byte_size);
+    if (!entry_.ParseFromCodedStream(&coded_input)) {
+      fprintf(stderr, "error reading input stream\n");
+      return nullptr;
+    }
+    auto tuple = std::make_unique<souffle::RamDomain[]>(kEntryElements);
```

Comment:
`souffle::mk` rather than absl::make_unique?

Reply:
Turns out souffle::Own is an alias for std::unique_ptr (my guess is they migrated from their own thing), but consistency is good.

Reply:
Actually, scratch that--souffle::mk doesn't work here. They use make_unique for the same purpose (e.g. https://github.com/souffle-lang/souffle/blob/e1fffe4ef91421cbf975c6898700949156cf5605/src/include/souffle/io/ReadStreamJSON.h#L105).